lgc/patch/NggPrimShader.cpp

```diff
@@ -1756,9 +1735,24 @@ void NggPrimShader::constructPrimShaderWithGs(Module *module) {
     m_nggFactor.primShaderTableAddrHigh = primShaderTableAddrHigh;
 
     // Record ES-GS vertex offsets info
-    m_nggFactor.esGsOffsets01 = esGsOffsets01;
-    m_nggFactor.esGsOffsets23 = esGsOffsets23;
-    m_nggFactor.esGsOffsets45 = esGsOffsets45;
+    auto esGsOffset0 = m_builder->CreateAnd(esGsOffsets01, 0xFFFF);
+    auto esGsOffset1 = m_builder->CreateLShr(esGsOffsets01, 16);
+    esGsOffset1 = m_builder->CreateAnd(esGsOffset1, 0xFFFF);
```

Comment:
I think it is better to move the three lines into a small function. Something like "void getLoHiOfInt32(Value *, Value **, Value **)"? Then you can pass in the addresses of m_nggFactor.esGsOffsetX as the arguments.

Reply:
OK. I will add a helper function to handle this.

Reply:
I add a helper CreateUBfe to do bitfield extraction because it is a special case to get low part and high part from a 32-bit integer. Actually, we need more general uses in this source file and we have several code pieces to replace amdgcn_ubfe (will make a future commit to do this).

Reply:
sounds ok to me.